server/util/rexec/rexec.go

```diff
@@ -0,0 +1,452 @@
+	gstatus "google.golang.org/grpc/status"
+)
+
+const (
+	defaultDigestFunction = repb.DigestFunction_SHA256
+)
+
+type state int
+
+const (
+	initial state = iota
+	uploading
+	starting
+	waiting
+)
+
+// Cmd provides a handle on a remote command.
+// It is not safe for concurrent use.
+type Cmd struct {
+	// Arguments are the executable and arguments for the command.
+	Arguments []string
+	// Dir is a local dir to be used as the action's root directory. The current
+	// directory contents will be uploaded from there, and outputs will be
+	// written there.
+	//
+	// Unlike exec.Cmd, an empty string here means "do not upload inputs or
+	// download outputs", rather than "use the current working directory".
+	Dir string
+	// Env is a slice of "NAME=VALUE" pairs to be used as environment variables.
+	// Unlike exec.Cmd, a nil value here means "use an empty env" instead of
+	// "use the current command's environment".
+	Env []string
+
+	// InstanceName is the remote instance name.
+	InstanceName string
+	// Platform is a slice of "NAME=VALUE" pairs to be used as platform
+	// properties.
+	Platform []string
+	// DigestFunction is the digest function to use for CAS transfers (defaults
+	// to SHA256).
+	DigestFunction repb.DigestFunction_Value
+	// Timeout is the timeout to be applied to the remote command. Note that the
+	// context timeout used for execution only applies to the RPC. Unless this
+	// timeout is also specified, the remote execution may continue beyond the
+	// local context deadline.
+	Timeout time.Duration
+
+	env           environment.Env
+	state         state
+	actionDigest  *repb.Digest
+	stream        *RetryingStream
+	response      *repb.ExecuteResponse
+	waitErr       error
+	commandResult *interfaces.CommandResult
+}
+
+// Command returns a new Cmd from the given executable and arguments.
+func Command(env environment.Env, executable string, args ...string) *Cmd {
+	return &Cmd{
+		env:       env,
+		Arguments: append([]string{executable}, args...),
+	}
+}
+
+// Close closes any streams established to the server.
+func (c *Cmd) Close() error {
+	if c.stream != nil {
+		return c.stream.CloseSend()
+	}
+	return nil
+}
+
+func (c *Cmd) String() string {
+	return fmt.Sprintf("%s", c.Arguments)
+}
+
+// Upload transfers all of the data to cache that is needed to run the action.
+// This includes the Action and Command protos, as well as the input tree.
+func (c *Cmd) Upload(ctx context.Context) error {
+	if c.state >= uploading {
+		return nil
+	}
+	c.state = uploading
+
+	p, err := AssemblePlatform(c.Platform...)
+	if err != nil {
+		return err
+	}
+	envVars, err := AssembleEnvironmentVariables(c.Env...)
+	if err != nil {
+		return err
+	}
+	cmd := &repb.Command{
+		Arguments:            c.Arguments,
+		EnvironmentVariables: envVars,
+		Platform:             p,
+	}
+	var commandDigest, inputRootDigest *repb.Digest
+	eg, egctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		ctx := egctx
```

Comment:
nit: prolly clearer if you juse use gctx directly in the calls rather than redefining ctx like this?

Reply:
Done